### PR TITLE
feat: add static code diagnostic check-for-equals-in-render-object-setters

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

- [ ] Documentation update
- [ ] Bug fix
- [X] New rule
- [ ] Changes an existing rule
- [ ] Add autofixing to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

#997

### CODE REVIEW

**Comment by incendial** [comment]
@fzyzcjy if you have some time to take a look and test the rule - that'd very much appreciated!

**Comment by fzyzcjy** [comment]
Sure, let me have a look :)

As for testing, may not have time now, but you can run it on the `flutter/flutter` repo, and even possibly fire a bug :)

File: test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    _page = value;
+    markNeedsLayout();
+  }
+
+  double get overflowSpacing => _overflowSpacing;
+  double _overflowSpacing;
+  set overflowSpacing(double value) {
+    assert(value != null);
+    if (_overflowSpacing == value) return;
```
Comment (fzyzcjy) [inline test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart:20]: IIRC there is also a case:

```dart
if (_overflowSpacing != value) {
  _overflowSpacing = value;
  markNeedsLayout()
}
```

But forget whether this is rare or not. Maybe we can run it on flutter framework repo first and see

**Comment by fzyzcjy** [comment]
LGTM especially on the demo in the tests. But not sure how many false positives will happen in the real codebase - may need to run it actually :)

File: test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart (hunk 1 above)
Comment (incendial) [reply test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart:20]: Hm, thanks, looks valid

**Comment by incendial** [comment]
Mostly false-positives with the case you pointed out before, but those:
* https://github.com/flutter/flutter/blob/7fba37848c9f3007795be2d67a3207067b7a0894/packages/flutter/lib/src/rendering/platform_view.dart#L323
* https://github.com/flutter/flutter/blob/8b3aabd61dfd7b7b525ae4be04093ba507d1455a/packages/flutter/lib/src/rendering/editable.dart#L596
* https://github.com/flutter/flutter/blob/a7fe235c8cd1ff64018443b0cca9a19a751642a6/packages/flutter/lib/src/rendering/paragraph.dart#L132
* https://github.com/flutter/flutter/blob/d3bc2bbcdd814beb343719d9b7fed58b3807675f/packages/flutter/lib/src/material/range_slider.dart#L891

look interesting. WDYT?

**Comment by fzyzcjy** [comment]
> https://github.com/flutter/flutter/blob/a7fe235c8cd1ff64018443b0cca9a19a751642a6/packages/flutter/lib/src/rendering/paragraph.dart#L132

No problem since it "return" when "idential"

Checking the rest 3

**Comment by fzyzcjy** [comment]
> https://github.com/flutter/flutter/blob/7fba37848c9f3007795be2d67a3207067b7a0894/packages/flutter/lib/src/rendering/platform_view.dart#L323

I strongly suspect this is a bug! That is what I have said in the issue title - "maybe even apply this linter to check potential bugs in Flutter source code"!

**Comment by incendial** [comment]
> What about giving a PR to flutter repo fixing this 🎉

Do you want to do it? It's mostly your effort finding this 🙂

**Comment by fzyzcjy** [comment]
> Do you want to do it?

Sure 🙂

**Comment by incendial** [comment]
@fzyzcjy found another place https://github.com/flutter/flutter/blob/2aa348b9407e96ffe4eca8e8f213c7984afad3f7/packages/flutter/lib/src/widgets/sliver_prototype_extent_list.dart#L135, WDYT?

**Comment by fzyzcjy** [comment]
@incendial Not very sure. The only interesting call is:

image

However it is named "insertRenderObjectChild", where by saying "insert" I suspect it will not happen unconditionally on every update. So I guess it is normal.

File: test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart (hunk 1 above)
Comment (incendial) [reply test/src/analyzers/lint_analyzer/rules/rules_list/check_for_equals_in_render_object_setters/examples/example.dart:20]: Fixed, ~10 false-positives left, but I think they're mostly for manual ignore

**Pull request merged**